### PR TITLE
feat: Add warning for using CPU mode

### LINKED ISSUE #1215: [TAB-432] Add latency warning in terminal when launching Tabby server with CPU device
**Please describe the feature you want**

Running Tabby on CPU devices normally experience significant latency, usually 10s+). We want to add a warning when launching Tabby server to set the right expectation to users.

**Additional context**

Example code generates terminal warning with `eprintln`

[https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/tabby-common/src/usage.rs#L24](https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/tabby-common/src/usage.rs#L24)

<sub>From [SyncLinear.com](https://synclinear.com) | [TAB-432](https://linear.app/tabbyml/issue/TAB-432/add-latency-warning-in-terminal-when-launching-tabby-server-with-cpu)</sub>

### PR DESCRIPTION
**Pull request by boxbeam**
Closes #1215

### CODE REVIEW

**Comment by wsxiaoys** [comment]
please use eprintln and ANSI color code, as suggested in #1215

**Comment by boxbeam** [comment]
> please use eprintln and ANSI color code, as suggested in #1215

I'd used `warn` because it prints between other logger outputs, which looks kind of strange. I'll include screenshots both ways so you can decide.

image

image

File: crates/tabby/src/services/model.rs [hunk 1]
```diff
@@ -65,6 +68,16 @@ impl PromptInfo {
 }
 
 fn create_ggml_engine(device: &Device, model_path: &str, parallelism: u8) -> impl TextGeneration {
+    if !device.ggml_use_gpu() {
+        tabby_common::show_info(
+            "LATENCY",
```
Comment (wsxiaoys) [inline crates/tabby/src/services/model.rs:73]:
```suggestion
            "CPU Device",
```

File: crates/tabby/src/services/model.rs [hunk 2]
```diff
@@ -65,6 +68,16 @@ impl PromptInfo {
...
+            HeaderFormat::BoldBlue,
+            &[
+                "Tabby is running on CPU! Completions will be very slow.",
+                "If you have a supported GPU, please pass --device [cuda/rocm/metal]",
```
Comment (wsxiaoys) [inline crates/tabby/src/services/model.rs:77]:
```suggestion
                "Tabby is currently running on the CPU. Completions may be slow, but it should suffice for testing purposes.",
                "For better performance, considering deploy Tabby on GPU devices",
```

File: crates/tabby-common/src/lib.rs [hunk 3]
```diff
@@ -70,3 +70,29 @@ pub struct Package {
...
+    }
+}
+
+pub fn show_info(header: &str, style: HeaderFormat, content: &[&str]) {
+    eprintln!("  {}", style.format(header));
+    for line in content {
+        eprintln!("  {line}");
+    }
+    eprintln!();
+}
```
Comment (wsxiaoys) [inline crates/tabby-common/src/lib.rs:98]: extract to `terminal.lib`

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: crates/tabby-common/src/lib.rs [hunk 4]
```diff
@@ -70,3 +70,29 @@ pub struct Package {
...
+            HeaderFormat::BoldBlue => "\x1b[34;1m",
+        }
+    }
+
+    fn format(&self, header: &str) -> String {
+        format!("{}{header}\x1b[0m", self.prefix())
+    }
+}
+
+pub fn show_info(header: &str, style: HeaderFormat, content: &[&str]) {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/lib.rs:92]: please also refactor https://github.com/TabbyML/tabby/blob/d4c375e42fc611fd6784f3b64e33f87caf4f6328/ee/tabby-webserver/src/schema/auth.rs#L46

**Pull request merged**